Optimize configuration infrastructure code

mwperina opened the pull request:
- Use constants for option names
- Introduce common configuration infrastructure
- Adapt agent configuration to the new infrastructure
- Adapt manager configuration to the new infrastructure
- Remove env vars handling from log infra
- Simplify logging initialization
- Remove the old logging infrastructure
- Rename config option Name to NodeName
- Rename config option Host to ManagerHost
- Rename config option Port to ManagerPort
- Rename config option Nodes to AllowedNodeNames
- Rename config option Level to LogLevel
- Rename config option Target to LogTarget
- Rename config option Quiet to LogIsQuiet
- Make config files more aligned with systemd config files
- Optimize configuration files

Fixes https://github.com/containers/hirte/issues/172

mwperina committed:
Use constants for option names

Introduce `libhirte/common/config.h` which should be the main focal
point for application configuration. In this patch only option names are
defined there and reused in all places, where configuration is being
parsed.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Introduce common configuration infrastructure

Introduces common configuration infracture, which provides:

* initialization and disposal of a global configuration
* loading of configuration from a specified configuration file
* setting and getting a value of a specified configuration option

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Adapt agent configuration to the new infrastructure

Adapt agent code around configuration to use the new configuration
infrastructure from `libhirte/common/config.h`.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Adapt manager configuration to the new infrastructure

Adapt manager code around configuration to use the new configuration
infrastructure from `libhirte/common/config.h`.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Remove the old logging infrastructure

Removes the old logging infrastructure from `libhirte/ini/config.h`.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Remove env vars handling from log infra

Removes environment variables handling from logging infrastructure,
because environment variables are now propagated to configuration, so
logging can be initialized only from configuration.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Simplify logging initialization

Reduce number of methods invocation to initialize logging settings from
loaded configuration.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Name to NodeName

Renames config option Name to NodeName to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Host to ManagerHost

Renames config option Host to ManagerHost to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Port to ManagerPort

Renames config option Port to ManagerPort to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Nodes to AllowedNodeNames

Renames config option Nodes to AllowedNodeNames to make this option
more clear in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Level to LogLevel

Renames config option Level to LogLevel to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Target to LogTarget

Renames config option Target to LogTarget to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Rename config option Quiet to LogIsQuiet

Renames config option Quiet to LogIsQuiet to make this option more clear
in configuration files.

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Make config files more aligned with systemd config files

Systemd configuration file format is described at
https://www.freedesktop.org/software/systemd/man/systemd.syntax.html
Change in the patch:

1. Renames `Manager` section to `hirte`
2. Renames `Node` section to `hirte-agent`
3. Removes `Logging` section as logging options can be used in both
   `hirte` and `hirte-agent` sections
2. Uses '#' character for comments instead of ';'

Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Optimize configuration files

1. Adds `hirte-default.conf`, which will contain all existing
   options with their default options and which should be in
   subsequent patches loaded as a default configuration for hirte
   and hirte-agent.

2. Add basic description to both `hirte.conf` and `agent.conf`

3. Mention only mandatory options in `hirte.conf` and `agent.conf`
   which needs to be set for each hirte/hirte-agent installation.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on config/hirte-agent.conf:
```diff
@@ -1,9 +1,16 @@
+
+#
+# The host used by hirte-agent to connect to hirte manager. It must be a valid
+# FQDN, IPv4 or IPv6 address.
```
```suggestion
# IPv4 or IPv6 address.
```
I think currently `hirte` only supports IPs.

mwperina replied:
OK, removing for now

engelmi commented on config/hirte-default.conf:
```diff
@@ -0,0 +1,50 @@
+#
+# A comma separated list of unique hirte-agent names. These are used as
+# a whitelist to validate any register request from each hirte-agent.
+VALID_NODE_NAMES=
```
```suggestion
EXPECTED_NODES=
```
For the whole system to be running, `hirte` requires those nodes to be present (which is currently not enforced) and expects them to be there. Valid would mean its ok if some are missing.

mwperina replied:
Yeah, but I'm not sure `EXPECTED` tells the whole story, so how about `ALLOWED_NODE_NAMES` or `APPROVED_NODE_NAMES`?

engelmi replied:
True. `ALLOWED_NODE_NAMES` sounds good :)

mwperina replied:
Changed

engelmi commented on src/agent/agent.c:
```diff
@@ -369,44 +370,61 @@ bool agent_set_name(Agent *agent, const char *name) {
-        if (config == NULL) {
+        result = config_initialize();
+        if (result != 0) {
+                fprintf(stderr, "Error initializing configuration, error code '%d'.", result);
```
```suggestion
                fprintf(stderr, "Error initializing configuration: '%s'.", strerror(-result));
```
With `strerror` we get the string for the code, e.g. `Out of memory`

mwperina replied:
Done

engelmi commented on src/agent/agent.c:
```diff
@@ -369,44 +370,61 @@ bool agent_set_name(Agent *agent, const char *name) {
+                fprintf(stderr,
+                        "Error loading configuration '%s', error code '%d'.",
+                        CFG_ETC_HIRTE_AGENT_CONF,
+                        result);
```
```suggestion
                fprintf(stderr,
                        "Error loading configuration '%s': %s",
                        CFG_ETC_HIRTE_AGENT_CONF,
                        strerror(-result));
```

mwperina replied:
Done

engelmi commented on src/agent/agent.c:
```diff
@@ -369,44 +370,61 @@ bool agent_set_name(Agent *agent, const char *name) {
-        topic = config_lookup_topic(config, "Node");
-        if (topic == NULL) {
-                return true;
+        // TODO: Do we really need to support custom config files passed as parameters from command line?
```
Good question. I'd prefer to have that option to be flexible. What do others think?

ygalblum replied:
Since the code is already there. I don't see a reason to remove it. I think we'll find it useful in tests

mwperina replied:
OK, removed the TODO comment

engelmi commented on src/agent/agent.c:
```diff
@@ -369,44 +370,61 @@ bool agent_set_name(Agent *agent, const char *name) {
+                        fprintf(stderr,
+                                "Error loading configuration file '%s', error code '%d'.",
+                                configfile,
+                                result);
```
```suggestion
                                "Error loading configuration file '%s': %s",
                                configfile,
                                strerror(-result));
```

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+}
+
+
+#pragma GCC diagnostic push
```
Instead of disabling the GCC diagnostic `unused parameters`, you can use the `UNUSED` 
So, in this case, it would be
```C
static int parsing_handler(UNUSED void *user, UNUSED const char *section, const char *name, const char *value) { // NOLINT
```
We use it [here](https://github.com/containers/hirte/blob/main/src/manager/manager.c#L632), for example - where the library requires all parameters, but we only use a part of it. 
Could you change it also for the other usages of `#pragma GCC diagnostic` in this file?

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+bool config_get_bool_value(const char *option_name) {
+        bool result = false;
+        const char *value = config_get_value(option_name);
+        if (value != NULL) {
```
There were implementations for `char* -> bool` in the (now deleted) `config.h`file:
https://github.com/containers/hirte/blob/main/src/libhirte/ini/config.c#L205-L221
I think the `is_true_value(...)` function should be reused here - it also accepts `t` and `y` as true values and uses the string cmp utility functions from [common/common.h](https://github.com/containers/hirte/blob/main/src/libhirte/common/common.h#L13-L17)

mwperina replied:
Fixed

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+                return found->value;
+        } else {
+                return NULL;
+        }
```
```suggestion
        if (found != NULL) {
                return found->value;
        }
        return NULL;
```

mwperina replied:
Fixed

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+#pragma GCC diagnostic pop
+
+int config_load_from_file(const char *config_file) {
+        assert(config_file != NULL);
```
```suggestion
        if (config_file == NULL) {
                return -EINVAL;
        }
```
I think in this case we could just return an error code and let the caller handle it. Could you change this also in all other places?

mwperina replied:
Fixed

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+int config_set_value(const char *option_name, const char *option_value) {
+        char *name_copy = strdup(option_name);
+        if (name_copy == NULL) {
+                return ENOMEM;
```
```suggestion
                return -ENOMEM;
```
I think in C its best practice to return negative error codes so the caller just does a `r<0` check and then inspects the code. For `strerror` (to get the string description of the code), it then gets turned into a positive one again.

mwperina replied:
Fixed

engelmi commented on src/manager/manager.c:
```diff
@@ -265,35 +265,49 @@ bool manager_set_port(Manager *manager, const char *port_s) {
-        if (config == NULL) {
+        result = config_initialize();
+        if (result != 0) {
+                fprintf(stderr, "Error initializing configuration, error code '%d'.", result);
```
```suggestion
                fprintf(stderr, "Error initializing configuration: %s", strerror(-result));
```

engelmi commented on src/manager/manager.c:
```diff
@@ -265,35 +265,49 @@ bool manager_set_port(Manager *manager, const char *port_s) {
+                        CFG_ETC_HIRTE_CONF,
+                        NULL); // TODO: https://github.com/containers/hirte/issues/148
+        if (result != 0) {
+                fprintf(stderr, "Error loading configuration '%s', error code '%d'.", CFG_ETC_HIRTE_CONF, result);
```
```suggestion
                fprintf(stderr, "Error loading configuration '%s': %s", CFG_ETC_HIRTE_CONF, strerror(-result));
```

mwperina replied:
Done

engelmi commented on src/manager/manager.c:
```diff
@@ -265,35 +265,49 @@ bool manager_set_port(Manager *manager, const char *port_s) {
+                        fprintf(stderr,
+                                "Error loading configuration file '%s', error code '%d'.",
+                                configfile,
+                                result);
```
```suggestion
                        fprintf(stderr,
                                "Error loading configuration file '%s': %s",
                                configfile,
                                strerror(-result));
```

mwperina replied:
Done

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+
+        if (default_config_file != NULL) {
+                result = config_load_from_file(default_config_file);
+                if (result != 0) {
```
Should starting `hirte` really fail if one of these are missing? For example, it will fail if `/etc/hirte.conf` is missing even if I'd override its config by a custom one (e.g. from the cli option). For production this might be ok, but for developing/testing its a bit harder to use. 
Maybe after loading config a kind of validation should be done, e.g. is config port really set?

mwperina replied:
OK makes sense, users might delete /etc/hirte*.conf, but we should not fail if they are not there.

But the default config file should be in `/usr/share/{hirte|hirte-agent}/config/hirte-default.conf` so it's mandatory part of RPM. If user removes such file, we should fail.

engelmi commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,175 @@
+/*
+ * The application configuration object.
+ */
+struct hashmap *config;
```
```suggestion
static struct hashmap *config;
```

mwperina replied:
Done

engelmi commented on src/manager/manager.c:
```diff
@@ -265,35 +265,49 @@ bool manager_set_port(Manager *manager, const char *port_s) {
+                                "Error loading configuration file '%s', error code '%d'.",
+                                configfile,
+                                result);
+                        config_dispose();
```
I think `config_dispose()` needs to be called also for the other fail-case returns in this functions (e.g. when `manager_set_port` fails). Instead of this, how about calling `config_dispose()` on the callers side? But then we'd have the same issue in `main.c`. Maybe the config should be part of the manger struct and the `config_dispose` should be called only once when the manager is freed. What do you think?

mwperina replied:
Hmm, not sure where to handle it. It there a way to create some hook reacting on KILL or TERM signal? Registering a function reacting on this signal might be a good way how to free objects, which should live the whole application lifetime

engelmi replied:
Yes, we have the `service/shutdown.c` where we also add callbacks to the signals:
https://github.com/containers/hirte/blob/main/src/libhirte/service/shutdown.c#L88-L132
These signals are setup when starting the manager and ensures that while running a SIGTERM all memory is freed properly. 

So I think the lifetime of this `hashmap` should be linked to the [manager struct](https://github.com/containers/hirte/blob/main/src/manager/manager.h#L10-L31). So in the [manager_new](https://github.com/containers/hirte/blob/main/src/manager/manager.c#L19-L48) where we could allocate it, e.g.
```C
// ...
manager->config = config_new(); // maybe rename from config_initialize and return the hashmap instead of an int
```
and in [manager_unref](https://github.com/containers/hirte/blob/main/src/manager/manager.c#L55-L100) we can call free(hashmap).

mwperina replied:
OK, I've added invocation of `config_dispose()` to both `manager_unref()` and `agent_unref()`

ygalblum commented on src/agent/agent.c:
```diff
@@ -370,44 +374,58 @@ bool agent_set_name(Agent *agent, const char *name) {
-        hirte_log_init_from_config(config);
-        // overwrite config settings with env vars
-        hirte_log_init_from_env();
+        result = config_load_complete_cofiguration(
```
typo: missing `n` in `configuration`

mwperina replied:
Fixed

ygalblum commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,180 @@
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+int option_compare(const void *a, const void *b, void *udata) { // NOLINT
```
Shouldn't these two functions be `static`?

mwperina replied:
Makes sense, fixed

ygalblum commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,180 @@
+int config_load_from_env() {
+        const char *env_vars[] = { ENV_LOG_LEVEL, ENV_LOG_TARGET, ENV_LOG_IS_QUIET };
+
+        const char *option_names[] = { CFG_LOG_LEVEL, CFG_LOG_TARGET, CFG_LOG_IS_QUIET };
```
I probably missed it, but just making sure - these environment variables are documented, right?

engelmi replied:
Yes, like the configuration keys for the config files, they are documented in the recently added man pages (e.g. for hirte: https://github.com/containers/hirte/blob/main/doc/man/hirte.1.md#environment-variables).

ygalblum commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,180 @@
+                        // empty string mean non-existent value, so store NULL
+                        value_copy = NULL;
+                }
+        }
```
I don't like this cast. Not sure why the compiler allowed it. I think this should be better:

```suggestion
        char *value_copy = NULL;
        if (option_value != NULL && strlen(option_value) > 0) {
            // storing non NULL value, so copy is needed
            value_copy = strdup(option_value);
            if (value_copy == NULL) {
                free(name_copy);
                return ENOMEM;
            }
        }
```

mwperina replied:
Fixed

ygalblum commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,180 @@
+
+void config_iterate(bool (*iter)(const void *item, void *udata), void *udata) {
+        hashmap_scan(config, iter, udata);
+}
```
Make sure to add an empty newline at the end of the file

mwperina replied:
Hmm, not sure why linter didn't scream on it :-(
Fixed

ygalblum commented on src/libhirte/common/config.h:
```diff
@@ -0,0 +1,120 @@
+/*
+ * Existing configuration environment options
+ */
+#define ENV_LOG_LEVEL "HIRTE_LOG_LEVEL"
```
Should these consts be exported? Are they not meant to be used only inside the config.c file? 
Furthermore, since they are basically a concatenation of `HIRTE_` and the name of the `CFG` flag, consider not defining them directly and instead use a concatenation macro (https://gcc.gnu.org/onlinedocs/cpp/Concatenation.html). This will make sure the names are aligned and will make it simpler to support the such flags

mwperina replied:
Well, they are used for testing in config_test.c, but they are not meant to be used in other places, so not sure if we really to use above concatenation macros suggestions

ygalblum replied:
I'm not a fan of using consts and macros from the code in the test code as it can hide issues. For example, a typo here will not show in the test.

mwperina replied:
Sorry, I'm probably missing there something. Could you please give an example how concatenation macros could make this part more maintainable and readable?

ygalblum replied:
My bad, it wasn't concatenation macros that did the trick, just usual macros.

If we agree that the tests should not use the macros from the code, then they are only used in one place. Therefore, instead of defining them, they can be created using macros and
in the `config.c` file's `cfg_load_from_env` you  will not need to maintain two arrays:
```
#define EXTERAL_PREFIX "HIRTE_"
#define ENV_FLAG_NAME(option_name) ( EXTERAL_PREFIX option_name)

int cfg_load_from_env(struct hashmap *config) {
        const char *option_names[] = { CFG_LOG_LEVEL, CFG_LOG_TARGET, CFG_LOG_IS_QUIET };

        int length = sizeof(option_names) / sizeof(option_names[0]);
        for (int i = 0; i < length; i++) {
                char *value = getenv(ENV_FLAG(option_names[i]);
                if (value != NULL && strlen(value) > 0) {
                        // environment variable is defined and has a non-empty value, so store it to the configuration
                        int result = cfg_set_value(config, option_names[i], value);
                        if (result != 0) {
                                return result;
                        }
                }
        }
        return 0;
}
``` 
  The nice thing here is that to add an support for a new external variable you just need to edit the `option_names` array

mwperina replied:
With the differences between the current naming of configuration options and environment variables, it's no longer possible, but I've used the idea of having existing environment variable and relevant configuration options in arrays, so it's easy to add some new environment variable mapping

ygalblum commented on doc/man/hirte-agent.conf.5.md:
```diff
@@ -10,29 +10,21 @@ The basic file definition used to bootstrap hirte-agent.
 ## Format
 
-The hirte-agent configuration file is using the .ini file format.
+The hirte-agent configuration file is using the shell configuration file format.
```
Is this change still correct?

mwperina replied:
Changed to systemd configuration files format: https://www.freedesktop.org/software/systemd/man/systemd.syntax.html

pypingou commented on config/hirte-default.conf:
```diff
@@ -0,0 +1,50 @@
+
+#
+# A comma separated list of unique hirte-agent names. These are used as
+# a whitelist to validate any register request from each hirte-agent.
```
```suggestion
# to validate any register request from each hirte-agent.
```

pypingou replied:
let's avoid `whitelist`

mwperina replied:
Done

alexlarsson commented on src/libhirte/common/config.h:
```diff
@@ -12,3 +23,120 @@
+ *
+ * Returns 0 if successful, otherwise standard error code
+ */
+int cfg_initialize(struct hashmap **config);
```
I would probably add a custom struct to wrap the hashmap, like:

In header: 
```
typedef struct config config;
config *cfg_new(void);
void cfg_free(config *config)
```

And then in the C file:
```
struct config {
  struct hashmap *;
};
```
This way you hide the hashmap details from the users.

alexlarsson replied:
And make it easy to later add more fields.

mwperina replied:
Done

mwperina committed:
Rename existing `config` struct to `_config`

Renames existing `config` struct to `_config`, so we can use `config`
struct for the new implementation introduced in follow-up patches.

Signed-off-by: Martin Perina <mperina@redhat.com>

Yarboa commented on .github/workflows/e2e.yaml:
```diff
@@ -52,10 +52,10 @@ jobs:
           # Change node name in config file
           podman exec -it hirte-agent-bar \
-          sed -i "s/\(Name=\)\(.*\)/\1bar/" /hirte/example-hirte-agent.ini
+          sed -i "s/\(NODE_NAME=\)\(.*\)/\1bar/" /hirte/example-hirte-agent.ini
```
I would advise for better test e2e coverage should be done through config files and not the ini.

ini files could be tested as part of code unitests
At the end hirte should run as a service on each node, isnt it?

mwperina replied:
This will be solved in some of the follow up PRs which will rewrite e2e to use RPM and then update regular config files such as `/etc/hirte-agent.conf`

Yarboa reviewed (changes requested):
Is the doc/example files? are *.ini or *.conf

Do we really need the -c option in the CLI?

alexlarsson commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,249 @@
+                // error during hashmap initialization
+                return -ENOMEM;
+        }
+        (*config)->default_section = strdup(CFG_SECT_GLOBAL);
```
Should check for NULL here.

mwperina replied:
Done

alexlarsson commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,249 @@
+#pragma GCC diagnostic pop
+
+int cfg_initialize(struct config **config) {
+        *config = malloc(sizeof(struct config));
```
Use a local variable for config so you can assign *config only at the end when everything is successful.

mwperina replied:
Done

alexlarsson commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,249 @@
+int cfg_initialize(struct config **config) {
+        *config = malloc(sizeof(struct config));
+        if (config == NULL) {
+                // error during configuration structure initialization
```
We should free the partially initialized struct on failure

alexlarsson replied:
Sorry, this affects the error branches below only.

mwperina replied:
Done

alexlarsson commented on src/agent/agent.c:
```diff
@@ -391,44 +394,70 @@ bool agent_set_name(Agent *agent, const char *name) {
+                }
+        }
+
+        result = cfg_set_default_section(agent->config, CFG_SECT_AGENT);
```
Shouldn't this be set before loading the files?

alexlarsson replied:
I guess that doesn't really matter.

mwperina replied:
This is used only in `cfg_get_value` and `cfg_set_value`, which are not called during parsing files or environment, but it needs to be called before accessing existing configuration

alexlarsson commented on config/agent/hirte-default.conf:
```diff
@@ -0,0 +1,31 @@
+#
+# The unique name of this agent.
+# It's mandatory to set this option for each hirte-agent.
+NODE_NAME=
```
Caps, should be NodeName

mwperina replied:
Sorry, forgot to change the default file added in the last commit, too many refactorings :-(
Fixed

alexlarsson commented on config/hirte/hirte-default.conf:
```diff
@@ -0,0 +1,25 @@
+#
+# The port the manager listens on to establish connections with the hirte
+# agents.
+MANAGER_PORT=2020
```
Also, all caps

mwperina replied:
Fixed

alexlarsson commented on src/libhirte/common/config.c:
```diff
@@ -0,0 +1,259 @@
+        char *section_copy = strdup(CFG_SECT_GLOBAL);
+        if (section_copy == NULL) {
+                // error during default section initialization
+                free(new_cfg->cfg_store);
```
should be hashmap_free(), not free(). Although, optionally you can call cfg_dispose() instead, if you make sure cfg_dispose handles some of the fields being NULL.

mwperina replied:
Right, missed that :-(
Calling hashmap_free makes more sense to me in this use case.

pypingou commented on config/agent/agent.conf:
```diff
@@ -0,0 +1,17 @@
+NodeName=
+
+#
+# The host used by hirte-agent to connect to hirte manager. It must be a valid
```
```suggestion
# The IP address that hirte-agent can use to connect to hirte manager. It must be a valid
```

mwperina replied:
Fixed

pypingou commented on config/agent/hirte-default.conf:
```diff
@@ -0,0 +1,31 @@
+NodeName=
+
+#
+# The host used by hirte-agent to connect to hirte manager. It must be a valid
```
```suggestion
# The IP address that hirte-agent can use to connect to hirte manager. It must be a valid
```

mwperina replied:
Fixed

pypingou commented on config/hirte/hirte-default.conf:
```diff
@@ -0,0 +1,25 @@
+
+#
+# A comma separated list of unique hirte-agent names. Only nodes with names
+# mentioned in the list can connect to hirte manager.
```
```suggestion
# mentioned in the list can connect to hirte manager. These names are defined in the agent's configuration file under `NodeName`.
```

mwperina replied:
Done
